spec/props_template_spec.rb

```diff
@@ -296,4 +296,64 @@
       }.to raise_error(Props::InvalidScopeForArrayError)
     end
   end
+
+  context "partial!" do
+    it "renders with a partial" do
+      json = render(<<~PROPS)
+        json.partial! partial: 'simple'
+      PROPS
+
+      expect(json).to eql_json({
+        foo: "bar"
+      })
+    end
+
+    it "renders with locale" do
+      json = render(<<~PROPS)
+        json.partial! partial: 'simple', locale: :de
+      PROPS
+
+      expect(json).to eql_json({
+        foo: "Kein"
+      })
+    end
+
+    it "renders with variants" do
+      json = render(<<~PROPS)
+        json.partial! partial: 'simple', variants: :grid
+      PROPS
+
+      expect(json).to eql_json({
+        foo: "Grid"
+      })
+    end
+
+    it "renders with a partial with locals" do
+      json = render(<<~PROPS)
+        json.partial! partial: 'profile', locals: {email: 'joe@joe.com'}
+      PROPS
+
+      expect(json).to eql_json({
+        email: "joe@joe.com"
+      })
+    end
+
+    it "renders an array of partials" do
+      json = render(<<~PROPS)
+        emails = [
+          'joe@j.com',
+          'foo@f.com',
+        ]
+
+        json.array! emails do |email|
+          json.partial! partial: 'profile', locals: {email: email}
+        end
+      PROPS
+
+      expect(json).to eql_json([
+        {email: "joe@j.com"},
+        {email: "foo@f.com"}
+      ])
+    end
+  end
```

Comment:
I'll continue to review this, but as a tip, we never want to `json.partial!` inside an `json.array!`. This would slow the views down considerably as each partial may result in a separate view lookup (i believe its less impactful on production). 

props_template's partial implementation (specified with a key) is optimized for this case, and more (russian doll caching)[https://github.com/thoughtbot/props_template?tab=readme-ov-file#caching].

Reply:
this is just test for checking such rendering opportunity, and array! with partial option is better for real rendering